backend/inspirehep/orcid/domain_models.py

```diff
@@ -125,42 +125,56 @@ def push(self):
+                    self._push_work_with_clashing_identifier()
+                    putcode = self._post_or_put_work(putcode)
+                except orcid_client_exceptions.WorkAlreadyExistsException:
+                    # If the putcode wasn't found, but we obtained
+                    # WorkAlreadyExistsException (e.g. records have the same doi)
+                    # it means that the external identifiers are duplicated
+                    # and we can't do anything about it.
                     raise exceptions.DuplicatedExternalIdentifierPusherException
-                self._push_work_with_clashing_identifier()
-                putcode = self._post_or_put_work(putcode)
-            except orcid_client_exceptions.WorkAlreadyExistsException:
-                putcode = self._cache_all_author_putcodes()
-                if not putcode:
-                    msg = (
-                        "No putcode was found in ORCID API for orcid={} and recid={}."
-                        " And the POST has previously failed for the same recid because"
-                        " the work had already existed".format(self.orcid, self.recid)
-                    )
-                    raise exceptions.PutcodeNotFoundInOrcidException(msg)
-                putcode = self._post_or_put_work(putcode)
+            else:
+                self._post_or_put_work(putcode)
         except orcid_client_exceptions.DuplicatedExternalIdentifierException:
             # We PUT a record changing its identifier, but there is another work
             # in ORCID with the same identifier. We need to find out the recid
             # of the clashing work in ORCID and push a fresh version of that
             # record.
             # This scenario might be triggered by a merge of 2 records in Inspire.
-            if self.do_fail_if_duplicated_identifier:
+            if self.pushing_duplicated_identifier:
                 raise exceptions.DuplicatedExternalIdentifierPusherException
             self._push_work_with_clashing_identifier()
             putcode = self._post_or_put_work(putcode)
         except orcid_client_exceptions.PutcodeNotFoundPutException:
+            # We try to push the work with invalid putcode, so we delete
+            # its putcode and push it without any putcode. If it turns out that the
+            # record already exists in ORCID we call the push method on
+            # it once again (without a putcode this time)
+            # to resolve WorkAlreadyExistsException.
             self.cache.delete_work_putcode()
             self.converter = OrcidConverter(
                 record=self.inspire_record,
                 url_pattern=current_app.config["LEGACY_RECORD_URL_PATTERN"],
                 put_code=None,
             )
-            putcode = self._post_or_put_work()
+            try:
+                putcode = self._post_or_put_work()
+            except orcid_client_exceptions.WorkAlreadyExistsException:
+                self.push()
```

Comment:
Here you might end in a loop again. What about doing simply
```suggestion
            putcode = self._cache_all_author_putcodes()
            putcode = self._post_or_put_work(putcode)
```
(and update the comment in this branch)?